File: time_converter.py
#!/usr/bin/env python3

import sys
import time
import base64
import struct
from datetime import datetime, timedelta

def milliseconds_to_days_and_time_since_midnight(millis):
    """
    Convert milliseconds since epoch to:
    1. Number of days since 1970
    2. Milliseconds since midnight

    Returns a tuple of (days_since_epoch, millis_since_midnight)
    """
    # Convert to seconds for datetime operations
    seconds = millis / 1000

    # Create datetime object
    dt = datetime.fromtimestamp(seconds)

    # Calculate days since epoch (1970-01-01)
    epoch = datetime(1970, 1, 1)
    days_since_epoch = (dt.date() - epoch.date()).days

    # Calculate milliseconds since midnight
    midnight = datetime(dt.year, dt.month, dt.day)
    seconds_since_midnight = (dt - midnight).total_seconds()
    millis_since_midnight = int(seconds_since_midnight * 1000)

    return (days_since_epoch, millis_since_midnight)

def compress_millis_base64(millis):
    """
    Compress milliseconds into a base64 string.
    This packs the milliseconds as a 64-bit integer, then encodes to base64.
    """
    # Pack as a 64-bit integer (8 bytes)
    packed = struct.pack('>Q', millis)
    # Convert to base64
    compressed = base64.b64encode(packed).decode('ascii')
    return compressed

def compress_millis_custom(millis):
    """
    Compress milliseconds into a custom alphanumeric format.
    Uses base36 encoding (0-9, a-z) for more compact representation than base64.
    """
    chars = "0123456789abcdefghijklmnopqrstuvwxyz"
    base = len(chars)
    result = ""

    # Convert to base36
    number = millis
    while number > 0:
        result = chars[number % base] + result
        number //= base

    return result or "0"

def compress_millis_base62(millis):
    """
    Compress milliseconds into a base62 format (0-9, a-z, A-Z).
    This is the most compact alphanumeric representation possible
    where case sensitivity is preserved (lowercase != uppercase).
    """
    chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    base = len(chars)  # 62
    result = ""

    # Convert to base62
    number = millis
    while number > 0:
        result = chars[number % base] + result
        number //= base

    return result or "0"

def compress_split_days_time(millis):
    """
    Compress timestamp by splitting into days since epoch and milliseconds since midnight,
    then encode each component separately using base62.
    """
    chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    base = len(chars)  # 62

    # Split into days since epoch and milliseconds since midnight
    days, millis_since_midnight = milliseconds_to_days_and_time_since_midnight(millis)

    # Convert days to base62
    days_result = ""
    days_number = days
    while days_number > 0:
        days_result = chars[days_number % base] + days_result
        days_number //= base

    # Ensure we have at least one character for days
    if not days_result:
        days_result = "0"

    # Convert milliseconds since midnight to base62
    time_result = ""
    time_number = millis_since_midnight
    while time_number > 0:
        time_result = chars[time_number % base] + time_result
        time_number //= base

    # Ensure we have at least one character for time
    if not time_result:
        time_result = "0"

    # Add a separator between the two components - using ":" as it's readable and common for time
    return days_result + ":" + time_result

def get_size_info(original, representation):
    """Calculate size information about a representation."""
    original_bytes = len(str(original))
    representation_bytes = len(representation)
    saving = original_bytes - representation_bytes
    saving_percent = (saving / original_bytes) * 100 if original_bytes > 0 else 0

    if saving >= 0:
        efficiency = f"Saved {saving} characters ({saving_percent:.2f}%)"
    else:
        efficiency = f"Increased by {abs(saving)} characters ({abs(saving_percent):.2f}%)"

    return {
        "original_size": original_bytes,
        "representation_size": representation_bytes,
        "efficiency": efficiency
    }

def process_timestamp(millis):
    """Process a timestamp and print all representations."""
    print(f"Original timestamp: {millis} milliseconds since epoch")
    print("-" * 50)

    # Format 1: Days since epoch and milliseconds since midnight
    days, millis_since_midnight = milliseconds_to_days_and_time_since_midnight(millis)
    days_repr = f"Days since epoch: {days}, Milliseconds since midnight: {millis_since_midnight}"
    print(days_repr)
    days_info = get_size_info(millis, f"{days},{millis_since_midnight}")
    print(f"Size: {days_info['representation_size']} chars, {days_info['efficiency']}")
    print("-" * 50)

    # Format 2: Base64 compression
    base64_repr = compress_millis_base64(millis)
    print(f"Base64 representation: {base64_repr}")
    base64_info = get_size_info(millis, base64_repr)
    print(f"Size: {base64_info['representation_size']} chars, {base64_info['efficiency']}")
    print("-" * 50)

    # Format 3: Custom base36 compression
    base36_repr = compress_millis_custom(millis)
    print(f"Base36 representation: {base36_repr}")
    base36_info = get_size_info(millis, base36_repr)
    print(f"Size: {base36_info['representation_size']} chars, {base36_info['efficiency']}")
    print("-" * 50)

    # Format 4: Case-sensitive base62 compression (most compact alphanumeric)
    base62_repr = compress_millis_base62(millis)
    print(f"Base62 representation: {base62_repr}")
    base62_info = get_size_info(millis, base62_repr)
    print(f"Size: {base62_info['representation_size']} chars, {base62_info['efficiency']}")
    print("-" * 50)

    # Format 5: Split days/time with base62 encoding for each part
    split_base62_repr = compress_split_days_time(millis)
    print(f"Split Days/Time (Base62): {split_base62_repr}")
    split_base62_info = get_size_info(millis, split_base62_repr)
    print(f"Size: {split_base62_info['representation_size']} chars, {split_base62_info['efficiency']}")
    print("-" * 50)

    # Format 6: Hexadecimal representation
    hex_repr = format(millis, 'x')
    print(f"Hexadecimal representation: {hex_repr}")
    hex_info = get_size_info(millis, hex_repr)
    print(f"Size: {hex_info['representation_size']} chars, {hex_info['efficiency']}")
    print("-" * 50)

    # Summary
    print("Summary of representations:")
    print(f"1. Original:          {millis} ({len(str(millis))} chars)")
    print(f"2. Days/Millis:       {days},{millis_since_midnight} ({days_info['representation_size']} chars)")
    print(f"3. Base64:            {base64_repr} ({base64_info['representation_size']} chars)")
    print(f"4. Base36:            {base36_repr} ({base36_info['representation_size']} chars)")
    print(f"5. Base62:            {base62_repr} ({base62_info['representation_size']} chars)")
    print(f"6. Split Days/Time:   {split_base62_repr} ({split_base62_info['representation_size']} chars)")
    print(f"7. Hexadecimal:       {hex_repr} ({hex_info['representation_size']} chars)")

def main():
    if len(sys.argv) > 1:
        try:
            millis = int(sys.argv[1])
            process_timestamp(millis)
        except ValueError:
            print("Error: Please provide a valid integer for milliseconds.")
            sys.exit(1)
    else:
        # Use current time if no argument provided
        current_millis = int(time.time() * 1000)
        print("No timestamp provided. Using current time.")
        process_timestamp(current_millis)

if __name__ == "__main__":
    main()

File: time_converter.js
#!/usr/bin/env node

/**
 * Time Compression Utility
 *
 * This script converts millisecond timestamps to different formats:
 * 1. Number of days since 1970 and milliseconds since midnight
 * 2. Compressed formats using alphanumerical characters
 * 3. Shows different representations with information about format and size
 */

/**
 * Convert milliseconds since epoch to days since 1970 and milliseconds since midnight
 * @param {number} millis - milliseconds since epoch
 * @returns {object} Object containing days since epoch and milliseconds since midnight
 */
function millisecondsToDaysAndTime(millis) {
  // Create date object from milliseconds
  const date = new Date(millis);

  // Calculate days since epoch (Jan 1, 1970)
  const epoch = new Date(0); // 0 = Jan 1, 1970
  const daysSinceEpoch = Math.floor((date - epoch) / (24 * 60 * 60 * 1000));

  // Calculate milliseconds since midnight
  const midnight = new Date(
    date.getFullYear(),
    date.getMonth(),
    date.getDate()
  );
  const millisSinceMidnight = date - midnight;

  return {
    daysSinceEpoch,
    millisSinceMidnight
  };
}

/**
 * Compress milliseconds into a base64 string
 * @param {number} millis - milliseconds since epoch
 * @returns {string} Base64 encoded string
 */
function compressMillisBase64(millis) {
  // Convert to a buffer (8 bytes for a 64-bit integer)
  const buffer = Buffer.alloc(8);

  // Write the number as a 64-bit big-endian integer
  buffer.writeBigInt64BE(BigInt(millis), 0);

  // Convert to base64
  return buffer.toString('base64');
}

/**
 * Compress milliseconds into a custom base36 format (0-9, a-z)
 * @param {number} millis - milliseconds since epoch
 * @returns {string} Base36 encoded string
 */
function compressMillisCustom(millis) {
  const chars = "0123456789abcdefghijklmnopqrstuvwxyz";
  const base = chars.length;

  // Convert to Base36
  let result = "";
  let number = millis;

  do {
    result = chars[number % base] + result;
    number = Math.floor(number / base);
  } while (number > 0);

  return result;
}

/**
 * Compress milliseconds into a base62 format (0-9, a-z, A-Z)
 * This is the most compact alphanumeric representation possible
 * where case sensitivity is preserved (lowercase != uppercase)
 * @param {number} millis - milliseconds since epoch
 * @returns {string} Base62 encoded string
 */
function compressMillisBase62(millis) {
  const chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
  const base = chars.length; // 62

  // Convert to Base62
  let result = "";
  let number = millis;

  do {
    result = chars[number % base] + result;
    number = Math.floor(number / base);
  } while (number > 0);

  return result;
}

/**
 * Compress timestamp by splitting into days since epoch and milliseconds since midnight,
 * then encode each component separately using base62
 * @param {number} millis - milliseconds since epoch
 * @returns {string} Compact base62 encoded string with day and time components
 */
function compressSplitDaysTime(millis) {
  const chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
  const base = chars.length; // 62

  // Split into days since epoch and milliseconds since midnight
  const { daysSinceEpoch, millisSinceMidnight } = millisecondsToDaysAndTime(millis);

  // Convert days to Base62
  let daysResult = "";
  let daysNumber = daysSinceEpoch;

  do {
    daysResult = chars[daysNumber % base] + daysResult;
    daysNumber = Math.floor(daysNumber / base);
  } while (daysNumber > 0);

  // Convert milliseconds since midnight to Base62
  let timeResult = "";
  let timeNumber = millisSinceMidnight;

  do {
    timeResult = chars[timeNumber % base] + timeResult;
    timeNumber = Math.floor(timeNumber / base);
  } while (timeNumber > 0);

  // Add a separator between the two components - using ":" as it's readable and common for time
  return daysResult + ":" + timeResult;
}

/**
 * Calculate size information about a representation
 * @param {number} original - Original milliseconds value
 * @param {string} representation - The encoded representation
 * @returns {object} Size and efficiency information
 */
function getSizeInfo(original, representation) {
  const originalBytes = original.toString().length;
  const representationBytes = representation.toString().length;
  const saving = originalBytes - representationBytes;
  const savingPercent = originalBytes > 0 ? (saving / originalBytes) * 100 : 0;

  let efficiency;
  if (saving >= 0) {
    efficiency = `Saved ${saving} characters (${savingPercent.toFixed(2)}%)`;
  } else {
    efficiency = `Increased by ${Math.abs(saving)} characters (${Math.abs(savingPercent).toFixed(2)}%)`;
  }

  return {
    originalSize: originalBytes,
    representationSize: representationBytes,
    efficiency
  };
}

/**
 * Process a timestamp and print all representations
 * @param {number} millis - milliseconds since epoch
 */
function processTimestamp(millis) {
  console.log(`Original timestamp: ${millis} milliseconds since epoch`);
  console.log("-".repeat(50));

  // Format 1: Days since epoch and milliseconds since midnight
  const { daysSinceEpoch, millisSinceMidnight } = millisecondsToDaysAndTime(millis);
  const daysRepr = `${daysSinceEpoch},${millisSinceMidnight}`;
  console.log(`Days since epoch: ${daysSinceEpoch}, Milliseconds since midnight: ${millisSinceMidnight}`);
  const daysInfo = getSizeInfo(millis, daysRepr);
  console.log(`Size: ${daysInfo.representationSize} chars, ${daysInfo.efficiency}`);
  console.log("-".repeat(50));

  // Format 2: Base64 compression
  const base64Repr = compressMillisBase64(millis);
  console.log(`Base64 representation: ${base64Repr}`);
  const base64Info = getSizeInfo(millis, base64Repr);
  console.log(`Size: ${base64Info.representationSize} chars, ${base64Info.efficiency}`);
  console.log("-".repeat(50));

  // Format 3: Custom base36 compression
  const base36Repr = compressMillisCustom(millis);
  console.log(`Base36 representation: ${base36Repr}`);
  const base36Info = getSizeInfo(millis, base36Repr);
  console.log(`Size: ${base36Info.representationSize} chars, ${base36Info.efficiency}`);
  console.log("-".repeat(50));

  // Format 4: Case-sensitive base62 compression (most compact alphanumeric)
  const base62Repr = compressMillisBase62(millis);
  console.log(`Base62 representation: ${base62Repr}`);
  const base62Info = getSizeInfo(millis, base62Repr);
  console.log(`Size: ${base62Info.representationSize} chars, ${base62Info.efficiency}`);
  console.log("-".repeat(50));

  // Format 5: Split days/time with base62 encoding for each part
  const splitBase62Repr = compressSplitDaysTime(millis);
  console.log(`Split Days/Time (Base62): ${splitBase62Repr}`);
  const splitBase62Info = getSizeInfo(millis, splitBase62Repr);
  console.log(`Size: ${splitBase62Info.representationSize} chars, ${splitBase62Info.efficiency}`);
  console.log("-".repeat(50));

  // Format 6: Hexadecimal representation
  const hexRepr = millis.toString(16);
  console.log(`Hexadecimal representation: ${hexRepr}`);
  const hexInfo = getSizeInfo(millis, hexRepr);
  console.log(`Size: ${hexInfo.representationSize} chars, ${hexInfo.efficiency}`);
  console.log("-".repeat(50));

  // Summary
  console.log("Summary of representations:");
  console.log(`1. Original:          ${millis} (${millis.toString().length} chars)`);
  console.log(`2. Days/Millis:       ${daysRepr} (${daysInfo.representationSize} chars)`);
  console.log(`3. Base64:            ${base64Repr} (${base64Info.representationSize} chars)`);
  console.log(`4. Base36:            ${base36Repr} (${base36Info.representationSize} chars)`);
  console.log(`5. Base62:            ${base62Repr} (${base62Info.representationSize} chars)`);
  console.log(`6. Split Days/Time:   ${splitBase62Repr} (${splitBase62Info.representationSize} chars)`);
  console.log(`7. Hexadecimal:       ${hexRepr} (${hexInfo.representationSize} chars)`);
}

/**
 * Main function to handle CLI arguments
 */
function main() {
  // Get milliseconds from command line arguments or use current time
  let millis;

  if (process.argv.length > 2) {
    millis = parseInt(process.argv[2], 10);
    if (isNaN(millis)) {
      console.error("Error: Please provide a valid integer for milliseconds.");
      process.exit(1);
    }
  } else {
    // Use current time if no argument provided
    millis = Date.now();
    console.log("No timestamp provided. Using current time.");
  }

  processTimestamp(millis);
}

// Run the main function
main();
